[Basic] Add isFile to FileSystem

aciidb0mb3r opened the pull request.

ddunbar commented:
/cc @abertelrud who had similar changes within #501

ddunbar commented on Sources/POSIX/stat.swift:
```diff
@@ -10,14 +10,34 @@
+        return check(libc.S_IFDIR)
+    }
+
+    private func check(_ mode: UInt16) -> Bool {
```
I think something like `hasMode(_ mode)` or `is(mode:` would read better.

ddunbar commented on Sources/POSIX/stat.swift:
```diff
@@ -10,14 +10,34 @@
 import libc
 
+/// A structure containing information about a file.
+public struct FileStatus {
```
I don't think `POSIX` should contain non-POSIX things, if we want to create wrappers, or higher level things I think they should go in `Basic`, but maybe stay private.

aciidb0mb3r replied:
I agree but I have mixed feelings about this one. POSIX is to make things "swifty" and platform independent which kinda fits here. But ok.

aciidb0mb3r commented:
@abertelrud Didn't see you had similar things in that PR. I don't mean to step on your toes, I just want a `isFile` on `FileSystem`. I can drop the wrapper commit and implement isFile via PathShim for now.

abertelrud commented:
@aciidb0mb3r: Please, go ahead and add what you need.  I had ended up putting a bunch of changes into a single PR, and Daniel appropriately pointed out that it would be better to have the changes as separate PRs that would be easier to review.  So I closed the PR and have been adding back the individual pieces as separate PRs.  The goal is to get rid of the last path-related functionality in Utility, and have a somewhat decent API in Basic until it all gets replaced with FileSystem later when we have more time.  But I hadn't gotten to adding this back yet (FileStat), so please go ahead if you want to.  You can pull code from my closed PR if it helps.  I do favor having a `kind` (or `type`, though that word is overloaded more than `kind` is) property and to compare that to enum values.  That property and enum also allows for better switch statements.  So while I favor comparing the kind to `.file` and `.directory` etc instead of `isFile()`, it's not ultimately that critical how we do that.

abertelrud commented:
Regarding POSIX:  my opinion is that rather than thinking of wrapping POSIX calls, we should implement an API that lets us express the intent we want to in the rest of the code, and then implement that API using whatever lower-level API is best.  I had gone the direction of using POSIX calls rather than Foundation since the latter didn't seem complete enough yet on Linux, and to avoid the overhead of going through Foundation which then just calls down to POSIX anyway, but ultimately what matters is the API we write the rest of SwiftPM in.  I would love it if Foundation had a Path type eventually (not just URL, since IMO the semantic differences between file system paths and random URLs are worth representing).  But until then, we should put into `Basics` whatever API we want to have the rest of SwiftPM use, and I would not personally be sad to see the `POSIX` module disappear completely.

aciidb0mb3r commented:
@abertelrud Thanks I added code from your PR (isFile, isDirectory and isSymlink) and moved the wrapper to `POSIX.swift` in Basic

@ddunbar @abertelrud re-review please

aciidb0mb3r commented:
@swift-ci please test

ddunbar commented on Sources/Basic/POSIX.swift:
```diff
@@ -0,0 +1,85 @@
+private func stat(_ path: AbsolutePath, followSymlink: Bool = true) throws -> libc.stat? {
+    var sbuf = libc.stat()
+    let rv = followSymlink ? stat(path.asString, &sbuf) : lstat(path.asString, &sbuf)
+    guard rv == 0 || errno == ENOENT else { throw SystemError.stat(errno, path.asString) }
```
This is a perfect example of why I don't think this is a good idea. Clients are calling `POSIX.stat` which we have lifted errno to `throws`, but we have decided here that `ENOENT` is a special case. However, that may not be appropriate for all clients, some clients might very reasonable expect this to map directly to `stat` and throw for every error return. That would lead to subtle bugs in otherwise correct looking `POSIX`-level code.

aciidb0mb3r replied:
I kept this method private because I wasn't sure what to do but I kept the methods which depend on stat throwing which can have same behaviour as you described. Should I make them non throwing till we figure out the best approach?

ddunbar replied:
My personal inclination currently is to first sort this out at the client level.

Currently `exists` doesn't throw, and I would prefer if we first migrate everything onto `FileSystem` and then figure out if it should throw. But of course that is also because I have a strong preference that it not throw.

I think making `POSIX.stat` itself throw makes sense.

ddunbar commented on Sources/POSIX/stat.swift:
```diff
@@ -10,14 +10,44 @@
+                 case S_IFCHR:  self = .chardev
+                 case S_IFSOCK: self = .socket
+             default:
+                 return nil
```
Shouldn't this be `unknown` instead of optional?

aciidb0mb3r replied:
I left `unknown` in the enum case by mistake. I like to indicate unknown by optionals more than a case for them. kind might exist but it won't be our wrapper `Kind`. What do you think?

ddunbar replied:
I'm fine with that, let's try it and if it feels unwieldy we can always revisit. We should ditch the unused case though.

abertelrud replied:
I actually quite strongly agree with Daniel here, which is why the kind was originally not optional and there was an `unknown`.  In the interest of landing this I'm OK doing that in a later PR though, but it feels wrong for some files to not have a kind.

ddunbar commented on Sources/POSIX/stat.swift:
```diff
@@ -10,14 +10,44 @@
+/// otherwise it is not (any symbolic links in path components other than the last one are always traversed).
+/// If symbolic links are followed and the file system entity at `path` is a symbolic link that points to a
+/// non-existent path, then this function returns nil.
+public func stat(_ path: String, followSymlink: Bool = true) throws -> libc.stat {
```
In keeping with my `POSIX` philosophy, I feel like `POSIX` should expose `stat` and `lstat` and `Basic` can have a version which adds the `followSymlink` semantics.

abertelrud replied:
Agreed.  The similar code in PR #501 was slated for `Basic` so the difference in name as actually a benefit.

aciidb0mb3r commented:
@ddunbar @abertelrud updated

abertelrud commented on Tests/Basic/XCTestManifests.swift:
```diff
@@ -30,6 +30,7 @@ public func allTests() -> [XCTestCaseEntry] {
         testCase(TemporaryFileTests.allTests),
         testCase(ThreadTests.allTests),
         testCase(WalkTests.allTests),
+        testCase(POSIXTests.allTests),
```
Nit: these are usually kept in alphabetical order.

aciidb0mb3r replied:
sure, updated.

abertelrud commented:
@aciidb0mb3r: Thanks!

ddunbar commented:
LGTM

ddunbar commented:
Will test and merge...

A couple follow ons I think we need:
1. Discussion of whether `exists` etc. should throw.
2. Move implementations into `FileSystem` and kill the free functions, slowly.
